sopel/deprecated.py

```diff
@@ -0,0 +1,178 @@
+    reason: Optional[str] = None,
+    version: Optional[str] = None,
+    removed_in: Optional[str] = None,
+    warning_in: Optional[str] = None,
+    stack_frame: int = -1,
+    func: Optional[Callable] = None,
+):
+    """Decorator to mark deprecated functions in Sopel's API
+
+    :param reason: optional text added to the deprecation warning
+    :param version: optional version number when the decorated function
+                        is deprecated
+    :param removed_in: optional version number when the deprecated function
+                           will be removed
+    :param warning_in: optional version number when the decorated function
+                           should start emitting a warning when called
+    :param stack_frame: optional stack frame to output; defaults to
+                            ``-1``; should almost always be negative
+    :param func: deprecated function
+    :return: a callable that depends on how the decorator is called; either
+             the decorated function, or a decorator with the appropriate
+             parameters
+
+    Any time the decorated ``func`` is called, a deprecation warning will be
+    printed to ``sys.stderr``, with the last frame of the traceback. The
+    optional ``warning_in`` argument suppresses the warning on Sopel versions
+    older than that, allowing for multi-stage deprecation timelines.
+
+    The decorator can be used with or without arguments::
+
+        from sopel.deprecated import deprecated
+
+        @deprecated
+        def func1():
+            print('func 1')
+
+        @deprecated()
+        def func2():
+            print('func 2')
+
+        @deprecated(reason='obsolete', version='7.0', removed_in='8.0')
+        def func3():
+            print('func 3')
+
+    which will output the following in a console::
+
+        >>> func1()
+        Deprecated: func1
+        File "<stdin>", line 1, in <module>
+        func 1
+        >>> func2()
+        Deprecated: func2
+        File "<stdin>", line 1, in <module>
+        func 2
+        >>> func3()
+        Deprecated since 7.0, will be removed in 8.0: obsolete
+        File "<stdin>", line 1, in <module>
+        func 3
+
+    The ``stack_frame`` argument can be used to choose which stack frame is
+    printed along with the message text. By default, this decorator prints the
+    most recent stack frame (the last entry in the list, ``-1``),
+    corresponding to where the decorated function itself was called. However,
+    in certain cases such as deprecating conditional behavior within an object
+    constructor, it can be useful to show a less recent stack frame instead.
+
+    .. note::
+
+        There is nothing that prevents this decorator to be used on a class's
+        method, or on any existing callable.
+
+    .. versionadded:: 7.0
+        Parameters ``reason``, ``version``, and ``removed_in``.
+
+    .. versionadded:: 7.1
+        The ``warning_in`` and ``stack_frame`` parameters.
+
```

Comment:
For total completeness, one could add this:

```suggestion

.. versionchanged:: 8.0
    Moved out of :mod:`sopel.tools` to solve circular dependency issues.

```